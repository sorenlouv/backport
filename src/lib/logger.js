const logger = {};
logger.log = (...args) => {
  console.log(...args);
};

logger.error = (...args) => {
  console.error(...args);
};

module.exports = logger;
