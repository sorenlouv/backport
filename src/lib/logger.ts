export function log(...args: unknown[]) {
  console.log(...args);
}

export function error(...args: unknown[]) {
  console.error(...args);
}
