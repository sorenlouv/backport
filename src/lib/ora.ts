import oraOriginal from 'ora';

/* eslint-disable @typescript-eslint/no-empty-function */
const oraMock = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  start: (text?: string) => oraMock,
  succeed: () => {},
  stop: () => {},
  fail: () => {},
  stopAndPersist: () => {},
  set text(value: string) {},
} as oraOriginal.Ora;

export function ora(
  interactive: boolean | undefined,
  text?: string | undefined
): oraOriginal.Ora {
  return interactive ? oraOriginal({ text }) : oraMock;
}

export type Ora = oraOriginal.Ora;
