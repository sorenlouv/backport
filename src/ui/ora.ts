import oraOriginal from 'ora';

/* eslint-disable @typescript-eslint/no-empty-function */
const oraMock = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  start: (text?: string) => oraMock,
  succeed: () => {},
  stop: () => {},
  fail: () => {},
  stopAndPersist: () => {},
  set text(value: string) {},
} as oraOriginal.Ora;

export function ora(
  ci: boolean | undefined,
  text?: string | undefined
): oraOriginal.Ora {
  return ci ? oraMock : oraOriginal({ text });
}
