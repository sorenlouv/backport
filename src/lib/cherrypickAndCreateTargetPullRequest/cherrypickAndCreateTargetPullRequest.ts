import chalk from 'chalk';
import { flatten } from 'lodash';
import { ValidConfigOptions } from '../../options/options';
import { getSourceBranchFromCommits } from '../getSourceBranchFromCommits';
import {
  createBackportBranch,
  deleteBackportBranch,
  pushBackportBranch,
  getRepoForkOwner,
} from '../git';
import { addAssigneesToPullRequest } from '../github/v3/addAssigneesToPullRequest';
import { addLabelsToPullRequest } from '../github/v3/addLabelsToPullRequest';
import { addReviewersToPullRequest } from '../github/v3/addReviewersToPullRequest';
import {
  createPullRequest,
  getTitle,
  getPullRequestBody,
  PullRequestPayload,
} from '../github/v3/createPullRequest';
import { validateTargetBranch } from '../github/v4/validateTargetBranch';
import { consoleLog } from '../logger';
import { sequentially } from '../sequentially';
import { Commit } from '../sourceCommit/parseSourceCommit';
import { autoMergeNowOrLater } from './autoMergeNowOrLater';
import { getBackportBranchName } from './getBackportBranchName';
import { getMergeCommits } from './getMergeCommit';
import { getTargetPRLabels } from './getTargetPRLabels';
import { waitForCherrypick } from './waitForCherrypick';

export async function cherrypickAndCreateTargetPullRequest({
  options,
  commits,
  targetBranch,
}: {
  options: ValidConfigOptions;
  commits: Commit[];
  targetBranch: string;
}): Promise<{ url: string; number: number; didUpdate: boolean }> {
  const backportBranch = getBackportBranchName(targetBranch, commits);
  const repoForkOwner = getRepoForkOwner(options);
  consoleLog(`\n${chalk.bold(`Backporting to ${targetBranch}:`)}`);

  await validateTargetBranch({ ...options, branchName: targetBranch });
  await createBackportBranch({
    options,
    sourceBranch: getSourceBranchFromCommits(commits),
    targetBranch,
    backportBranch,
  });

  const commitsFlattened = flatten(
    await Promise.all(commits.map((c) => getMergeCommits(options, c))),
  );

  await sequentially(commitsFlattened, (commit) =>
    waitForCherrypick(options, commit, targetBranch),
  );

  if (!options.dryRun) {
    await pushBackportBranch({ options, backportBranch });
    await deleteBackportBranch({ options, backportBranch });
  }

  const prPayload: PullRequestPayload = {
    owner: options.repoOwner,
    repo: options.repoName,
    title: getTitle({ options, commits, targetBranch }),
    body: getPullRequestBody({ options, commits, targetBranch }),
    head: `${repoForkOwner}:${backportBranch}`, // eg. sqren:backport/7.x/pr-75007
    base: targetBranch, // eg. 7.x
  };

  const targetPullRequest = await createPullRequest({ options, prPayload });

  // add assignees to target pull request
  const assignees = options.autoAssign
    ? [options.authenticatedUsername]
    : options.assignees;

  if (options.assignees.length > 0) {
    await addAssigneesToPullRequest({
      ...options,
      pullNumber: targetPullRequest.number,
      assignees,
    });
  }

  // add reviewers to target pull request
  if (options.reviewers.length > 0) {
    await addReviewersToPullRequest(
      options,
      targetPullRequest.number,
      options.reviewers,
    );
  }

  // add labels to target pull request
  if (options.targetPRLabels.length > 0) {
    const labels = getTargetPRLabels({
      interactive: options.interactive,
      targetPRLabels: options.targetPRLabels,
      commits,
      targetBranch,
    });
    await addLabelsToPullRequest({
      ...options,
      pullNumber: targetPullRequest.number,
      labels,
    });
  }

  // make PR auto mergable
  if (options.autoMerge) {
    await autoMergeNowOrLater(options, targetPullRequest.number);
  }

  // add labels to source pull requests
  if (options.sourcePRLabels.length > 0) {
    const promises = commits.map((commit) => {
      if (commit.sourcePullRequest) {
        return addLabelsToPullRequest({
          ...options,
          pullNumber: commit.sourcePullRequest.number,
          labels: options.sourcePRLabels,
        });
      }
    });

    await Promise.all(promises);
  }

  consoleLog(`View pull request: ${targetPullRequest.url}`);

  return targetPullRequest;
}
