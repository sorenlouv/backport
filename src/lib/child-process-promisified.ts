import childProcess from 'child_process';
import { promisify } from 'util';
import { logger } from './logger';
const execPromisified = promisify(childProcess.exec);

export async function exec(
  cmd: string,
  options: childProcess.ExecOptions & { cwd: string }
) {
  const res = await execPromisified(cmd, {
    maxBuffer: 100 * 1024 * 1024,
    ...options,

    // ensure that git commands return english error messages
    env: { ...process.env, LANG: 'en_US' },
  });

  return res;
}

export async function spawnPromise(
  cmd: string,
  cmdArgs: string[],
  cwd: string
): Promise<{
  cmdArgs: string[];
  code: number | null;
  stderr: string;
  stdout: string;
}> {
  const fullCmd = `${cmd} ${cmdArgs.join(' ')}`;
  logger.info(`Running command: "${fullCmd}"`);

  return new Promise(function (resolve, reject) {
    const subprocess = childProcess.spawn(cmd, cmdArgs, {
      cwd,

      // ensure that git commands return english error messages
      env: { ...process.env, LANG: 'en_US' },
    });
    let stderr = '';
    let stdout = '';

    subprocess.stdout.on('data', (data: string) => {
      stdout += data;
    });

    subprocess.stderr.on('data', (data: string) => {
      stderr += data;
    });

    subprocess.on('close', (code) => {
      if (code === 0 || code === null) {
        resolve({ cmdArgs, code, stderr, stdout });
      } else {
        const err = new SpawnError({ cmdArgs, code, stderr, stdout });
        logger.verbose(`Error when running command: "${fullCmd}"`, err);
        reject(err);
      }
    });

    subprocess.on('error', (err) => {
      reject(err);
    });
  });
}

export const spawnStream = (cmd: string, cmdArgs: ReadonlyArray<string>) => {
  return childProcess.spawn(cmd, cmdArgs, {
    env: { ...process.env, LANG: 'en_US' },
  });
};

export type SpawnErrorContext = {
  cmdArgs: string[];
  code: number;
  stderr: string;
  stdout: string;
};

export class SpawnError extends Error {
  context: SpawnErrorContext;
  constructor(context: SpawnErrorContext) {
    const cmdArgs = context.cmdArgs.join(' ');
    const message = `Code: ${
      context.code
    }, Args: "${cmdArgs}", Message: ${context.stderr.trim()}`;

    super(message);
    Error.captureStackTrace(this, SpawnError);
    this.name = 'SpawnError';
    this.message = message;
    this.context = context;
  }
}
