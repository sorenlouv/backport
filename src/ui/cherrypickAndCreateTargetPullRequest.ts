import chalk from 'chalk';
import dedent = require('dedent');
import isEmpty = require('lodash.isempty');
import ora = require('ora');
import { BackportOptions } from '../options/options';
import { HandledError } from '../services/HandledError';
import { exec } from '../services/child-process-promisified';
import { getRepoPath } from '../services/env';
import {
  cherrypick,
  createBackportBranch,
  deleteBackportBranch,
  pushBackportBranch,
  setCommitAuthor,
  getUnmergedFiles,
  addUnstagedFiles,
  finalizeCherrypick,
  getFilesWithConflicts,
} from '../services/git';
import { getShortSha } from '../services/github/commitFormatters';
import { addAssigneesToPullRequest } from '../services/github/v3/addAssigneesToPullRequest';
import { addLabelsToPullRequest } from '../services/github/v3/addLabelsToPullRequest';
import { createPullRequest } from '../services/github/v3/createPullRequest';
import { consoleLog, logger } from '../services/logger';
import { confirmPrompt } from '../services/prompts';
import { sequentially } from '../services/sequentially';
import { CommitSelected } from '../types/Commit';

export async function cherrypickAndCreateTargetPullRequest({
  options,
  commits,
  targetBranch,
}: {
  options: BackportOptions;
  commits: CommitSelected[];
  targetBranch: string;
}) {
  const backportBranch = getBackportBranch(targetBranch, commits);
  consoleLog(`\n${chalk.bold(`Backporting to ${targetBranch}:`)}`);

  await createBackportBranch({ options, targetBranch, backportBranch });
  await sequentially(commits, (commit) => waitForCherrypick(options, commit));

  if (options.resetAuthor) {
    await setCommitAuthor(options, options.username);
  }

  await pushBackportBranch({ options, backportBranch });
  await deleteBackportBranch({ options, backportBranch });

  const targetPullRequest = await createPullRequest({
    options,
    commits,
    targetBranch,
    backportBranch,
  });

  // add assignees to target pull request
  if (options.assignees.length > 0) {
    await addAssigneesToPullRequest(
      options,
      targetPullRequest.number,
      options.assignees
    );
  }

  // add labels to target pull request
  if (options.targetPRLabels.length > 0) {
    await addLabelsToPullRequest(
      options,
      targetPullRequest.number,
      options.targetPRLabels
    );
  }

  // add labels to source pull requests
  if (options.sourcePRLabels.length > 0) {
    const promises = commits.map((commit) => {
      if (commit.pullNumber) {
        return addLabelsToPullRequest(
          options,
          commit.pullNumber,
          options.sourcePRLabels
        );
      }
    });
    await Promise.all(promises);
  }

  consoleLog(`View pull request: ${targetPullRequest.html_url}`);

  return targetPullRequest;
}

/*
 * Returns the name of the backport brancha
 *
 * Examples:
 * For a single PR: `backport/7.x/pr-1234`
 * For a single commit: `backport/7.x/commit-abcdef`
 * For multiple: `backport/7.x/pr-1234_commit-abcdef`
 */
export function getBackportBranch(
  targetBranch: string,
  commits: CommitSelected[]
) {
  const refValues = commits
    .map((commit) =>
      commit.pullNumber
        ? `pr-${commit.pullNumber}`
        : `commit-${getShortSha(commit.sha)}`
    )
    .join('_')
    .slice(0, 200);
  return `backport/${targetBranch}/${refValues}`;
}

async function waitForCherrypick(
  options: BackportOptions,
  commit: CommitSelected
) {
  const spinnerText = `Cherry-picking: ${chalk.greenBright(
    commit.formattedMessage
  )}`;
  const cherrypickSpinner = ora(spinnerText).start();

  if (options.dryRun) {
    cherrypickSpinner.succeed(`Dry run: ${spinnerText}`);
    return;
  }

  try {
    const { needsResolving } = await cherrypick(options, commit);

    // no conflicts encountered
    if (!needsResolving) {
      cherrypickSpinner.succeed();
      return;
    }

    // cherrypick failed due to conflicts
    cherrypickSpinner.fail();
  } catch (e) {
    cherrypickSpinner.fail();
    throw e;
  }

  // resolve conflicts automatically
  if (options.autoFixConflicts) {
    const autoResolveSpinner = ora(
      'Attempting to resolve conflicts automatically'
    ).start();

    const filesWithConflicts = await getFilesWithConflicts(options);
    const repoPath = getRepoPath(options);
    const didAutoFix = await options.autoFixConflicts({
      files: filesWithConflicts,
      directory: repoPath,
      logger,
    });

    // conflicts were automatically resolved
    if (didAutoFix) {
      autoResolveSpinner.succeed();
      return;
    }
    autoResolveSpinner.fail();
  }

  /*
   * Commit could not be cleanly cherrypicked: Initiating conflict resolution
   */

  if (options.editor) {
    const repoPath = getRepoPath(options);
    await exec(`${options.editor} ${repoPath}`, {});
  }

  // list files with conflict markers and require user to resolve them
  await listConflictingFiles(options);

  // list unmerged files and require user to confirm adding+comitting them
  await listUnstagedFiles(options);

  // Conflicts resolved and unstaged files will now be staged and committed
  const stagingSpinner = ora(`Finalizing cherrypick`).start();
  try {
    // add unstaged files
    await addUnstagedFiles(options);

    // Run `git commit`
    await finalizeCherrypick(options);
    stagingSpinner.succeed();
  } catch (e) {
    stagingSpinner.fail();
    throw e;
  }
}

async function listConflictingFiles(options: BackportOptions) {
  const checkForConflicts = async (): Promise<void> => {
    const filesWithConflicts = await getFilesWithConflicts(options);

    if (isEmpty(filesWithConflicts)) {
      return;
    }

    consoleLog(''); // linebreak
    const res = await confirmPrompt(
      dedent(`
        ${chalk.reset(
          `The following files from ${getRepoPath(options)} have conflicts:`
        )}
        ${chalk.reset(
          filesWithConflicts.map((file) => ` - ${file}`).join('\n')
        )}

        ${chalk.reset.italic(
          'You do not need to `git add` or `git commit` the files - simply fix the conflicts.'
        )}

        Press ENTER when the conflicts are resolved
      `)
    );
    if (!res) {
      throw new HandledError('Aborted');
    }

    await checkForConflicts();
  };

  await checkForConflicts();
}

async function listUnstagedFiles(options: BackportOptions) {
  const unmergedFiles = await getUnmergedFiles(options);

  if (isEmpty(unmergedFiles)) {
    return;
  }

  consoleLog(''); // linebreak
  const res = await confirmPrompt(
    dedent(`
      ${chalk.reset(`The following files are unstaged:`)}
      ${chalk.reset(unmergedFiles.map((file) => ` - ${file}`).join('\n'))}



      Press ENTER to stage them
    `)
  );
  if (!res) {
    throw new HandledError('Aborted');
  }
  consoleLog(''); // linebreak
}
