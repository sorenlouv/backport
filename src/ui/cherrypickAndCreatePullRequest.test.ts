import nock from 'nock';
import ora from 'ora';
import { BackportOptions } from '../options/options';
import * as childProcess from '../services/child-process-promisified';
import * as logger from '../services/logger';
import * as prompts from '../services/prompts';
import { ExecError } from '../test/ExecError';
import { CommitSelected } from '../types/Commit';
import { PromiseReturnType } from '../types/PromiseReturnType';
import { SpyHelper } from '../types/SpyHelper';
import { cherrypickAndCreateTargetPullRequest } from './cherrypickAndCreateTargetPullRequest';

describe('cherrypickAndCreateTargetPullRequest', () => {
  let execSpy: SpyHelper<typeof childProcess.exec>;
  let addLabelsScope: ReturnType<typeof nock>;
  let consoleLogSpy: SpyHelper<typeof logger['consoleLog']>;

  beforeEach(() => {
    execSpy = jest
      .spyOn(childProcess, 'exec')

      // mock all exec commands to respond without errors
      .mockResolvedValue({ stdout: '', stderr: '' });

    consoleLogSpy = jest.spyOn(logger, 'consoleLog');

    // ensure labels are added
    addLabelsScope = nock('https://api.github.com')
      .post('/repos/elastic/kibana/issues/1337/labels', {
        labels: ['backport'],
      })
      .reply(200);
  });

  afterEach(() => {
    addLabelsScope.done();
    jest.clearAllMocks();
  });

  describe('when commit has a pull request reference', () => {
    let res: PromiseReturnType<typeof cherrypickAndCreateTargetPullRequest>;

    beforeEach(async () => {
      const options = {
        assignees: [] as string[],
        githubApiBaseUrlV3: 'https://api.github.com',
        fork: true,
        targetPRLabels: ['backport'],
        prDescription: 'myPrSuffix',
        prTitle: '[{targetBranch}] {commitMessages}',
        repoName: 'kibana',
        repoOwner: 'elastic',
        username: 'sqren',
        sourceBranch: 'myDefaultSourceBranch',
        sourcePRLabels: [] as string[],
      } as BackportOptions;

      const commits: CommitSelected[] = [
        {
          sourceBranch: '7.x',
          sha: 'mySha',
          formattedMessage: 'myCommitMessage (#1000)',
          pullNumber: 1000,
          targetBranchesFromLabels: [],
        },
        {
          sourceBranch: '7.x',
          sha: 'mySha2',
          formattedMessage: 'myOtherCommitMessage (#2000)',
          pullNumber: 2000,
          targetBranchesFromLabels: [],
        },
      ];

      const scope = nock('https://api.github.com')
        .post('/repos/elastic/kibana/pulls', {
          title: '[6.x] myCommitMessage (#1000) | myOtherCommitMessage (#2000)',
          head: 'sqren:backport/6.x/pr-1000_pr-2000',
          base: '6.x',
          body:
            'Backports the following commits to 6.x:\n - myCommitMessage (#1000)\n - myOtherCommitMessage (#2000)\n\nmyPrSuffix',
        })
        .reply(200, { number: 1337, html_url: 'myHtmlUrl' });

      res = await cherrypickAndCreateTargetPullRequest({
        options,
        commits,
        targetBranch: '6.x',
      });

      scope.done();
    });

    it('returns the expected response', () => {
      expect(res).toEqual({ html_url: 'myHtmlUrl', number: 1337 });
    });

    it('should make correct git commands', () => {
      expect(execSpy.mock.calls).toMatchSnapshot();
    });

    it('logs correctly', () => {
      expect(consoleLogSpy.mock.calls.length).toBe(2);
      expect(consoleLogSpy.mock.calls[0][0]).toMatchInlineSnapshot(`
        "
        Backporting to 6.x:"
      `);
      expect(consoleLogSpy.mock.calls[1][0]).toMatchInlineSnapshot(
        `"View pull request: myHtmlUrl"`
      );
    });

    it('should start the spinner with the correct text', () => {
      expect((ora as any).mock.calls.map((call: any) => call[0]))
        .toMatchInlineSnapshot(`
        Array [
          "Pulling latest changes",
          "Cherry-picking: myCommitMessage (#1000)",
          "Cherry-picking: myOtherCommitMessage (#2000)",
          "Pushing branch \\"sqren:backport/6.x/pr-1000_pr-2000\\"",
          undefined,
          "Creating pull request",
          "Adding labels: backport",
        ]
      `);
    });
  });

  describe('when commit does not have a pull request reference', () => {
    let res: PromiseReturnType<typeof cherrypickAndCreateTargetPullRequest>;
    beforeEach(async () => {
      const options = {
        assignees: [] as string[],
        githubApiBaseUrlV3: 'https://api.github.com',
        fork: true,
        targetPRLabels: ['backport'],
        prTitle: '[{targetBranch}] {commitMessages}',
        repoName: 'kibana',
        repoOwner: 'elastic',
        username: 'sqren',
        sourcePRLabels: [] as string[],
      } as BackportOptions;

      const scope = nock('https://api.github.com')
        .post('/repos/elastic/kibana/pulls', {
          title: '[6.x] myCommitMessage (mySha)',
          head: 'sqren:backport/6.x/commit-mySha',
          base: '6.x',
          body:
            'Backports the following commits to 6.x:\n - myCommitMessage (mySha)',
        })
        .reply(200, { number: 1337, html_url: 'myHtmlUrl' });

      res = await cherrypickAndCreateTargetPullRequest({
        options,
        commits: [
          {
            sourceBranch: '7.x',
            sha: 'mySha',
            formattedMessage: 'myCommitMessage (mySha)',
            targetBranchesFromLabels: [],
          },
        ],
        targetBranch: '6.x',
      });
      scope.done();
    });

    it('returns the expected response', () => {
      expect(res).toEqual({ html_url: 'myHtmlUrl', number: 1337 });
    });
  });

  describe('when cherry-picking fails', () => {
    let res: PromiseReturnType<typeof cherrypickAndCreateTargetPullRequest>;
    let promptSpy: SpyHelper<typeof prompts['confirmPrompt']>;
    let execSpy: ReturnType<typeof setupExecSpy>;
    beforeEach(async () => {
      // spies
      promptSpy = jest.spyOn(prompts, 'confirmPrompt').mockResolvedValue(true);
      execSpy = setupExecSpy();

      const options = {
        assignees: [] as string[],
        fork: true,
        targetPRLabels: ['backport'],
        prTitle: '[{targetBranch}] {commitMessages}',
        repoName: 'kibana',
        repoOwner: 'elastic',
        username: 'sqren',
        sourceBranch: 'myDefaultSourceBranch',
        sourcePRLabels: [] as string[],
      } as BackportOptions;

      const scope = nock('https://api.github.com')
        .post('/repos/elastic/kibana/pulls', {
          title: '[6.x] myCommitMessage',
          head: 'sqren:backport/6.x/commit-mySha',
          base: '6.x',
          body: 'Backports the following commits to 6.x:\n - myCommitMessage',
        })
        .reply(200, { html_url: 'myHtmlUrl', number: 1337 });

      res = await cherrypickAndCreateTargetPullRequest({
        options,
        commits: [
          {
            sourceBranch: '7.x',
            sha: 'mySha',
            formattedMessage: 'myCommitMessage',
            targetBranchesFromLabels: [],
          },
        ],
        targetBranch: '6.x',
      });

      scope.done();
    });

    it('creates pull request', () => {
      expect(res).toEqual({ html_url: 'myHtmlUrl', number: 1337 });
    });

    it('shows the right prompts', () => {
      expect(promptSpy.mock.calls.length).toBe(3);

      expect(promptSpy.mock.calls[0][0]).toMatchInlineSnapshot(`
        "Please fix the issues in: /myHomeDir/.backport/repositories/elastic/kibana

        Conflicting files:
         - /myHomeDir/.backport/repositories/elastic/kibana/conflicting-file.txt


        Press ENTER when the conflicts are resolved and files are staged"
      `);

      expect(promptSpy.mock.calls[1][0]).toMatchInlineSnapshot(`
        "Please fix the issues in: /myHomeDir/.backport/repositories/elastic/kibana

        Conflicting files:
         - /myHomeDir/.backport/repositories/elastic/kibana/conflicting-file.txt


        Press ENTER when the conflicts are resolved and files are staged"
      `);

      expect(promptSpy.mock.calls[2][0]).toMatchInlineSnapshot(`
        "Please fix the issues in: /myHomeDir/.backport/repositories/elastic/kibana


        Unstaged files:
         - /myHomeDir/.backport/repositories/elastic/kibana/conflicting-file.txt

        Press ENTER when the conflicts are resolved and files are staged"
      `);
    });

    it('calls exec correctly', () => {
      expect(execSpy.mock.calls).toMatchSnapshot();
    });

    it('calls ora correctly', () => {
      expect((ora as any).mock.calls.map((call: any) => call[0]))
        .toMatchInlineSnapshot(`
        Array [
          "Pulling latest changes",
          "Cherry-picking: myCommitMessage",
          "Finalizing cherrypick",
          "Pushing branch \\"sqren:backport/6.x/commit-mySha\\"",
          undefined,
          "Creating pull request",
          "Adding labels: backport",
        ]
      `);
    });

    it('logs correctly', async () => {
      expect(consoleLogSpy.mock.calls[0][0]).toMatchInlineSnapshot(`
        "
        Backporting to 6.x:"
      `);
      expect(consoleLogSpy.mock.calls[1][0]).toMatchInlineSnapshot(`
        "
        ----------------------------------------
        "
      `);
      expect(consoleLogSpy.mock.calls[2][0]).toMatchInlineSnapshot(`
        "
        ----------------------------------------
        "
      `);
      expect(consoleLogSpy.mock.calls[3][0]).toMatchInlineSnapshot(
        `"View pull request: myHtmlUrl"`
      );
    });
  });
});

function setupExecSpy() {
  let conflictingFilesCheck = 0;
  let unstagedFilesCheck = 0;
  return jest
    .spyOn(childProcess, 'exec')

    .mockImplementation(async (cmd) => {
      // createFeatureBranch
      if (cmd.includes('git checkout -B')) {
        return { stdout: 'create feature branch succeeded', stderr: '' };
      }

      // git fetch
      if (cmd.startsWith('git fetch')) {
        return { stderr: '', stdout: '' };
      }

      // cherrypick
      if (cmd === 'git cherry-pick mySha') {
        throw new ExecError({ cmd });
      }

      // getConflictingFiles
      if (cmd === 'git --no-pager diff --check') {
        conflictingFilesCheck++;
        if (conflictingFilesCheck >= 4) {
          return { stderr: '', stdout: '' };
        }

        throw new ExecError({
          code: 2,
          cmd,
          stdout: `conflicting-file.txt:1: leftover conflict marker\nconflicting-file.txt:3: leftover conflict marker\nconflicting-file.txt:5: leftover conflict marker\n`,
        });
      }

      // getUnstagedFiles
      if (cmd === 'git --no-pager diff --name-only') {
        unstagedFilesCheck++;
        if (unstagedFilesCheck >= 5) {
          return { stderr: '', stdout: '' };
        }
        return { stdout: `conflicting-file.txt\n`, stderr: '' };
      }

      // addUnstagedFiles
      if (cmd === 'git add --update') {
        return { stdout: ``, stderr: '' };
      }

      // finalizeCherrypick
      if (cmd.includes('git commit --no-edit')) {
        return { stdout: ``, stderr: '' };
      }

      // pushFeatureBranch
      if (cmd.startsWith('git push ')) {
        return { stdout: ``, stderr: '' };
      }

      // deleteFeatureBranch
      if (cmd.includes('git branch -D ')) {
        return { stdout: ``, stderr: '' };
      }

      throw new Error(`Missing mock for "${cmd}"`);
    });
}
