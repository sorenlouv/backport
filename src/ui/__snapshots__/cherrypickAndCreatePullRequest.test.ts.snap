// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cherrypickAndCreateTargetPullRequest when cherry-picking fails calls exec correctly 1`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git fetch elastic 6.x && git checkout -B backport/6.x/commit-mySha elastic/6.x --no-track",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch elastic 7.x:7.x --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --check",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --name-only",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --check",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --name-only",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --check",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --name-only",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --check",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --name-only",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --check",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git --no-pager diff --name-only",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git commit --no-edit",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git push sqren backport/6.x/commit-mySha:backport/6.x/commit-mySha --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git reset --hard && git checkout myDefaultSourceBranch && git branch -D backport/6.x/commit-mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
]
`;

exports[`cherrypickAndCreateTargetPullRequest when commit has a pull request reference should make correct git commands 1`] = `
Array [
  Array [
    "git reset --hard && git clean -d --force && git fetch elastic 6.x && git checkout -B backport/6.x/pr-1000_pr-2000 elastic/6.x --no-track",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch elastic 7.x:7.x --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git fetch elastic 7.x:7.x --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git cherry-pick mySha2",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git push sqren backport/6.x/pr-1000_pr-2000:backport/6.x/pr-1000_pr-2000 --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
  Array [
    "git reset --hard && git checkout myDefaultSourceBranch && git branch -D backport/6.x/pr-1000_pr-2000",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
    },
  ],
]
`;
